File: P41 First Missing Positive/Solution.java
class Solution {
    public int firstMissingPositive(int[] nums) {
        int n = nums.length;
        int[] num_occurrence = new int[n+1];
        for(int i = 0; i < n; i++) {
            if(nums[i] > 0 && nums[i] <= n) {
                num_occurrence[nums[i]] = 1;
            }
        }
        for(int i = 1; i < n + 1; i++) {
            if(num_occurrence[i] == 0) {
                return i;
            }
        }
        return n + 1;
    }

    public int firstMissingPositive2(int[] nums) {
        int i = 1;
        while(true) {
            if(!contains(nums, i)) {
                return i;
            }
            i++;
        }
    }
    
    private Boolean contains(int[] nums, int num) {
        for(int i = 0; i< nums.length; i++) {
            if(num == nums[i]) {
                return true;
            }
        }
        return false;
    }
}

File: P41 First Missing Positive/solution.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        n = len(nums)
        num_occurrence = [0] * (n + 1)
        for i in range(n):
            if nums[i] > 0 and nums[i] <= n:
                num_occurrence[nums[i]] = 1
        for i in range(1, n + 1):
            if num_occurrence[i] == 0:
                return i
        return n + 1
    
    def firstMissingPositive2(self, nums: List[int]) -> int:
        i = 1
        while True:
            if i not in nums:
                return i
            i+=1
        return i